src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+    DAWN_LOG(INFO) << "analyzing stencil " << stencilInstantiation->getName();
+    std::unordered_map<int, Extents> stencilDirtyFields;
+    stencilDirtyFields.clear();
+
+    for(const auto& multiStagePtr : stencil.getMultiStages()) {
+      MultiStage& multiStage = *multiStagePtr;
+      for(auto stageIt = multiStage.getStages().begin(); stageIt != multiStage.getStages().end();
+          ++stageIt) {
+        Stage& stage = (**stageIt);
+        for(const auto& domethod : stage.getDoMethods()) {
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
+              if(originalID == FieldType::FT_NotOriginal)
+                continue;
+              auto idWithExtents = dirtyFields.find(originalID);
+              if(idWithExtents == dirtyFields.end())
+                continue;
+              // If the access is horizontally pointwise or it is a horizontally cached field,
+              // we do not need to trigger a BC
+              if(readacccess.second.isHorizontalPointwise() ||
+                 stencilInstantiation->getCachedVariableSet().count(readacccess.first))
+                continue;
+              auto finder = allBCs.find(originalID);
+              // Check if a boundary condition for this variable was defined
+              if(finder == allBCs.end()) {
+                DAWN_ASSERT_MSG(
+                    false,
+                    dawn::format("In stencil %s we need a halo update on field %s but no "
+                                 "boundary condition is set.\nUpdate the stencil (outside the "
+                                 "do-method) with a boundary condition that calls a "
+                                 "stencil_function, e.g \n'boundary_condition(zero(), %s);'\n",
+                                 stencilInstantiation->getName(),
+                                 stencilInstantiation->getOriginalNameFromAccessID(originalID),
+                                 stencilInstantiation->getOriginalNameFromAccessID(originalID))
+                        .c_str());
+              }
+              // Calculate the extent and add it to the boundary-condition - Extent map
+              Extents fullExtents = calculateHaloExtents(
+                  stencilInstantiation->getNameFromAccessID(readacccess.first));
+              stencilInstantiation->getBoundaryConditionToExtentsMap().emplace(finder->second,
```

Comment:
use a set pattern rather than get + emplace. And maybe rename `finder`, since it is not very readable, it does not say what it contains